File: fragments/api/fastify/datastore/sqlite.ts
import { type FastifyInstance, type FastifyPluginOptions } from "fastify";
import {
  addItem,
  type CryptoCoin,
  deleteItem,
  getAllItems,
  getItemByTicker,
  getItemsAfterLaunchYear,
  updateItem,
} from "../../../sqlite_db/sqlite_crud";

export const routes = (fastify: FastifyInstance, _: FastifyPluginOptions) => {
  fastify.get<{
    Reply: {
      200: CryptoCoin[];
      500: { error: string };
    };
  }>("/coins", async (request, reply) => {
    try {
      const items = await getAllItems();
      return items;
    } catch (error) {
      request.log.error(error, "Error fetching coins");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.get<{
    Params: { ticker: string };
    Reply: {
      200: CryptoCoin;
      404: void;
      500: { error: string };
    };
  }>("/coins/:ticker", async (request, reply) => {
    try {
      const { ticker } = request.params;
      const item = await getItemByTicker(ticker.toUpperCase());

      if (!item) {
        return reply.code(404).send();
      }
      return item;
    } catch (error) {
      request.log.error(error, "Error fetching coin by ticker");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.get<{
    Params: { year: number };
    Reply: {
      200: CryptoCoin[];
      500: { error: string };
    };
  }>("/coins/after/:year", async (request, reply) => {
    try {
      const { year } = request.params;
      const items = await getItemsAfterLaunchYear(year);
      return items;
    } catch (error) {
      request.log.error(error, "Error fetching coins after launch year");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.put<{
    Body: Omit<CryptoCoin, "id" | "ticker">;
    Params: { ticker: string };
    Reply: {
      200: void;
      400: { error: string };
      500: { error: string };
    };
  }>("/coins/:ticker", {
    schema: {
      body: {
        type: "object",
        required: ["name", "launched"],
        properties: {
          name: { type: "string" },
          launched: { type: "number" },
        },
      },
    },
  }, async (request, reply) => {
    try {
      const { ticker } = request.params;
      const { name, launched } = request.body;
      await addItem({
        name,
        launched,
        ticker: ticker.toUpperCase(),
      });
      return;
    } catch (error) {
      request.log.error(error, "Error adding coin");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.delete<{
    Params: { ticker: string };
    Reply: {
      204: void;
      500: { error: string };
    };
  }>("/coins/:ticker", async (request, reply) => {
    try {
      const { ticker } = request.params;
      await deleteItem(ticker.toUpperCase());
      return reply.code(204).send();
    } catch (error) {
      request.log.error(error, "Error removing coin");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.patch<{
    Body: Omit<CryptoCoin, "id" | "ticker">;
    Params: { ticker: string };
    Reply: {
      200: CryptoCoin;
      404: void;
      500: { error: string };
    };
  }>("/coins/:ticker", {
    schema: {
      body: {
        type: "object",
        required: ["name", "launched"],
        properties: {
          name: { type: "string" },
          launched: { type: "number" },
        },
      },
    },
  }, async (request, reply) => {
    const { ticker } = request.params;
    const { name, launched } = request.body;
    try {
      const updatedItem = await updateItem({
        ticker: ticker.toUpperCase(),
        name,
        launched,
      });

      if (!updatedItem) {
        return reply.code(404).send();
      }
      return updatedItem;
    } catch (error) {
      request.log.error(error, "Error updating coin");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });
};

File: fragments/apis/gin/datastore/sqlite.go
package datastore

import (
	"database/sql"
	"errors"
	sqlite_crud "ff/sqlite_db"
	"fmt"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"
)

func SqliteRoutes(r *gin.RouterGroup) {
	coins := r.Group("/coins")

	coins.GET("", SqliteGetCoins)
	coins.GET("/:ticker", SqliteGetCoinByTicker)
	coins.GET("/after/:year", SqliteGetCoinsAfterYear)
	coins.PUT("/:ticker", SqliteAddCoin)
	coins.PATCH("/:ticker", SqliteUpdateCoin)
	coins.DELETE("/:ticker", SqliteDeleteCoin)
}

func SqliteGetCoins(c *gin.Context) {
	coins, err := sqlite_crud.GetAllItems()
	if err != nil {
		c.JSON(500, gin.H{"error": "Failed to retrieve coins"})
		return
	}
	c.JSON(200, coins)
}

func SqliteGetCoinByTicker(c *gin.Context) {
	ticker := strings.ToUpper(c.Param("ticker"))
	coin, err := sqlite_crud.GetItemByTicker(ticker)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			c.Status(404)
			return
		}
		c.JSON(500, gin.H{"error": fmt.Sprintf("Failed to get coin by ticker %s", ticker)})
		return
	}
	c.JSON(200, coin)
}

func SqliteGetCoinsAfterYear(c *gin.Context) {
	year, err := strconv.Atoi(c.Param("year"))
	if err != nil {
		c.JSON(400, gin.H{"error": fmt.Sprintf("Invalid year parameter %s", c.Param("year"))})
		return
	}
	coins, err := sqlite_crud.GetItemsAfterLaunchYear(year)
	if err != nil {
		c.JSON(500, gin.H{"error": fmt.Sprintf("Failed to get coins after year %d", year)})
		return
	}
	c.JSON(200, coins)
}

func SqliteAddCoin(c *gin.Context) {
	ticker := strings.ToUpper(c.Param("ticker"))
	var req CoinWriteRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(422, gin.H{"error": "Invalid payload"})
		return
	}
	coin := sqlite_crud.CryptoCoinWithoutId{Ticker: ticker, Name: req.Name, Launched: req.Launched}
	_, _, err := sqlite_crud.AddItem(coin)
	if err != nil {
		c.JSON(500, gin.H{"error": fmt.Sprintf("Failed to add coin with ticker %s", ticker)})
		return
	}
	c.Status(200)
}

func SqliteUpdateCoin(c *gin.Context) {
	ticker := strings.ToUpper(c.Param("ticker"))
	var req CoinWriteRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(422, gin.H{"error": "Invalid payload"})
		return
	}
	coin := sqlite_crud.CryptoCoinWithoutId{Ticker: ticker, Name: req.Name, Launched: req.Launched}
	updatedCoin, err := sqlite_crud.UpdateItem(coin)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			c.Status(404)
			return
		}
		c.JSON(500, gin.H{"error": fmt.Sprintf("Failed to update coin with ticker %s", ticker)})
		return
	}
	c.JSON(200, updatedCoin)
}

func SqliteDeleteCoin(c *gin.Context) {
	ticker := strings.ToUpper(c.Param("ticker"))
	_, err := sqlite_crud.DeleteItem(ticker)
	if err != nil {
		c.JSON(500, gin.H{"error": fmt.Sprintf("Failed to delete coin with ticker %s", ticker)})
		return
	}
	c.Status(204)
}
